modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,733 @@
+  return M;
+}
+
+
+/* Return a matrix (2D array) over domain ``Dom`` */
+proc Matrix(Dom, type eltType=real) where Dom.rank == 2 {
+  var M: [Dom] eltType;
+  return M;
+}
+
+
+/* Return a matrix (2D array) with domain and values of ``A`` */
+proc Matrix(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  var M: [Dom] eltType = A;
+  return M;
+}
+
+
+pragma "no doc"
+proc Matrix(Arrays...?n) {
+  type eltType = Arrays(1).eltType;
+  return Matrix((...Arrays), eltType=eltType);
+}
+
+
+/*
+    Return a matrix (2D array), given 2 or more vectors. Vectors are
+    concatenated such that the ``ith`` vector corresponds to the matrix slice:
+    ``A[i, ..]``
+
+    If `type` is omitted, it will be inferred from the first argument
+*/
+proc Matrix(const Arrays...?n, type eltType) {
+  // TODO -- assert all array domains are same length
+  //         Can this be done via type query?
+
+  if Arrays(1).domain.rank != 1 then compilerError("Matrix() expected 1D arrays");
+
+  const dim2 = 0..#Arrays(1).domain.dim(1).length,
+        dim1 = 0..#n;
+
+  var M: [{dim1, dim2}] eltType;
+
+  for i in dim1 do
+    M[i, ..] = Arrays(i+1)[..]: eltType;
+
+  return M;
+}
+
+
+/* Return an identity matrix over domain ``{0..#m, 0..#n}`` */
+proc eye(m, n, type eltType=real) {
+  var A: [{0..#m, 0..#n}] eltType;
+  const diagonal = min(m, n);
+  for i in 0..#diagonal do A[i, i] = 1: eltType;
+  return A;
+}
+
+
+/* Return an identity matrix over domain ``Dom`` */
+proc eye(Dom: domain(2), type eltType=real) {
+  var A: [Dom] eltType;
+  const diagonal = min(Dom.shape(1), Dom.shape(2));
+  return eye(Dom.shape(1), Dom.shape(2), eltType=eltType);
+}
+
+
+//
+// Matrix Operations
+//
+
+pragma "no doc"
+inline proc transpose(D: domain(1)) {
+  return D;
+}
+
+inline proc transpose(D: domain(2)) {
+  return {D.dim(2), D.dim(1)};
+}
+
+
+pragma "no doc"
+/* Transpose vector (no-op) */
+inline proc transpose(A: [?Dom]) where Dom.rank == 1 {
+  return A;
+}
+
+
+pragma "no doc"
+/* Syntactic sugar for transpose(Vector) */
+proc _array.T where this.domain.rank == 1 { return transpose(this); }
+
+
+/* Transpose vector or matrix.
+
+   .. note::
+
+      Since row vectors and columns vectors are indistinguishable, passing
+      a vector to this function will return that vector unchanged
+
+*/
+proc transpose(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  if !isBLASType(eltType) then
+    return _transpose(A);
+  else if Dom.shape(1) == 1 then
+    return reshape(A, transpose(Dom));
+  else if Dom.shape(2) == 1 then
+    return reshape(A, transpose(Dom));
+  else {
+    var B: [Dom] eltType = eye(Dom, eltType=eltType);
+    var C: [transpose(Dom)] eltType;
+    // TODO -- use native algorithm (BLAS not necessary or memory-efficient)
+    gemm(A, B, C, 1:eltType, 0:eltType, opA=Op.T);
+    return C;
+  }
+}
+
+
+/* Transpose vector or matrix */
+proc _array.T where this.domain.rank == 2 { return transpose(this); }
+
+
+/* Transpose matrix with native implementation */
```

Comment:
Really a TODO for a later time, but I think if you do some amount of tiling, you can improve cache performance here.  Why this isn't a part of BLAS is a mystery to me; I believe MKL has a routine that does do this (so you don't have to reinvent the wheel). Of course, if Chapel implements an array-level transpose as I suggested above, this would be moot.